3scale/apisonator
# THREESCALE-8989 ruby 3.0.X

### What

Support for Ruby 3.0.X

Implements [THREESCALE-8989](https://issues.redhat.com/browse/THREESCALE-8989)

CVEs Addressed:
* https://issues.redhat.com/browse/THREESCALE-8984

#### Dependency Updates 
Only first level dependencies updated shown here. Excluding testing deps also.
* [async 1.31.0](https://rubygems.org/gems/async/versions/1.31.0) (from 1.26.2)
* [async-redis 0.7.0](https://rubygems.org/gems/async-redis/versions/0.7.0) (from 0.5.1)
* [sinatra 2.2.4](https://rubygems.org/gems/sinatra/versions/2.2.4) (from 2.0.3)
* [sinatra-contrib 2.2.4](https://rubygems.org/gems/sinatra-contrib/versions/2.2.4) (from 2.0.3)
* [async-io 1.34.3](https://rubygems.org/gems/async-io/versions/1.34.3) (from 1.27.7)
* [redis-namespace 1.10.0](https://rubygems.org/gems/redis-namespace/versions/1.10.0) (from 1.8.0)

### Verification steps

Manual testing
* Build local image
```
docker build -t apisonator:runtime -f openshift/distro/ubi/8/Dockerfile .
```
* Run Redis
```
❯ docker run --net=host -ti --rm registry.redhat.io/rhscl/redis-5-rhel7:5 redis-server
```
* Run Listener
```
❯ docker run -it --rm --net=host --env CONFIG_REDIS_PROXY=127.0.0.1:6379 --env CONFIG_QUEUES_MASTER_NAME=127.0.0.1:6379/1 --env RACK_ENV=production --env CONFIG_INTERNAL_API_USER=internal_api_user --env CONFIG_INTERNAL_API_PASSWORD=internal_api_password --env PUMA_WORKERS=0 apisonator:runtime 3scale_backend start -e production -p 3000 -x /dev/stdout
```
* Run worker
```
❯ docker run -ti --rm --net=host --env CONFIG_REDIS_PROXY=127.0.0.1:6379/0 --env CONFIG_QUEUES_MASTER_NAME=127.0.0.1:6379/1 --env RACK_ENV=production --env CONFIG_WORKERS_LOGGER_FORMATTER=json apisonator:runtime bin/3scale_backend_worker run
```
* Populate some (fake) data
```
❯ docker run -it --rm --net=host quay.io/3scale/perftest-toolkit:buddhi-v1.1.0 -I http://127.0.0.1:3000 -B http://127.0.0.1:3000 -U internal_api_user -P internal_api_password -E http://echo-api.3scale.net -T onprem -A http://benchmark.3sca.net
```
* Get available URLS
```
❯ curl http://127.0.0.1:8089/paths/backend?lines=10
/transactions/authrep.xml?provider_key=29f5c9c6-b17b-4245-a12d-7eb52705ad8f&service_id=b92c4556-16c3-48a9-85c3-236060e13a56&app_id=8c0211817bddd50b&app_key=a3d8fe2a4e61d521&usage%5Ba5aa2153-6306-4053-b63f-26f1af4ec754%5D=1
/transactions/authrep.xml?provider_key=29f5c9c6-b17b-4245-a12d-7eb52705ad8f&service_id=b92c4556-16c3-48a9-85c3-236060e13a56&app_id=4a7011e7a47b8056&app_key=e33689d92c5034f6&usage%5Ba5aa2153-6306-4053-b63f-26f1af4ec754%5D=1
/transactions/authrep.xml?provider_key=29f5c9c6-b17b-4245-a12d-7eb52705ad8f&service_id=ad91cd17-6a24-476c-9435-3d14127d0787&app_id=daf8637eeaa14af6&app_key=ed1268d868e50474
/transactions/authrep.xml?provider_key=29f5c9c6-b17b-4245-a12d-7eb52705ad8f&service_id=b92c4556-16c3-48a9-85c3-236060e13a56&app_id=083f6d4caf710cb7&app_key=fff6b51cf57d4dc3
```
* Test few of the URLS, all of them should return `200 OK`
```
❯ curl -v "http://127.0.0.1:3000/transactions/authrep.xml?provider_key=29f5c9c6-b17b-4245-a12d-7eb52705ad8f&service_id=b92c4556-16c3-48a9-85c3-236060e13a56&app_id=8c0211817bddd50b&app_key=a3d8fe2a4e61d521&usage%5Ba5aa2153-6306-4053-b63f-26f1af4ec754%5D=1"

*   Trying 127.0.0.1:3000...
* Connected to 127.0.0.1 (127.0.0.1) port 3000 (#0)
> GET /transactions/authrep.xml?provider_key=29f5c9c6-b17b-4245-a12d-7eb52705ad8f&service_id=b92c4556-16c3-48a9-85c3-236060e13a56&app_id=8c0211817bddd50b&app_key=a3d8fe2a4e61d521&usage%5Ba5aa2153-6306-4053-b63f-26f1af4ec754%5D=1 HTTP/1.1
> Host: 127.0.0.1:3000
> User-Agent: curl/7.81.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/vnd.3scale-v2.0+xml
< Access-Control-Allow-Origin: *
< Access-Control-Expose-Headers: ETag, Link, 3scale-rejection-reason
< X-Content-Type-Options: nosniff
< Content-Length: 2224
< 
<?xml version="1.0" encoding="UTF-8"?><status><authorized>true</authorized><plan>83e35191-e62e-439d-bb8f-dc98ca947018</plan><usage_reports><usage_report metric="8ffefa9c-c3b8-4be9-9fbd-7904a2eeb416" period="day"><period_start>2023-04-19 00:00:00 +0000</period_start><period_end>2023-04-20 00:00:00 +0000</period_end><max_value>1000000000</max_value><current_value>0</current_value></usage_report><usage_report metric="8ffefa9c-c3b8-4be9-9fbd-7904a2eeb416" period="hour"><period_start>2023-04-19 09:00:00 +0000</period_start><period_end>2023-04-19 10:00:00 +0000</period_end><max_value>1000000000</max_value><current_value>0</current_value></usage_report><usage_report metric="a5aa2153-6306-4053-b63f-26f1af4ec754" period="day"><period_start>2023-04-19 00:00:00 +0000</period_start><period_end>2023-04-20 00:00:00 +0000</period_end><max_value>1000000000</max_value><current_value>1</current_value></usage_report><usage_report metric="a5aa2153-6306-4053-b63f-26f1af4ec754" period="hour"><period_start>2023-04-19 09:00:00 +0000</period_start><period_end>2023-04-19 10:00:00 +0000</period_end><max_value>1000000000</max_value><current_value>1</current_value></usage_report><usage_report metric="63b99e0d-da73-44dc-b412-d8c390472b50" period="day"><period_start>2023-04-19 00:00:00 +0000</period_start><period_end>2023-04-20 00:00:00 +0000</period_end><max_value>1000000000</max_value><current_value>1</current_value></usage_report><usage_report metric="63b99e0d-da73-44dc-b412-d8c390472b50" period="hour"><period_start>2023-04-19 09:00:00 +0000</period_start><period_end>2023-04-19 10:00:00 +0000</period_end><max_value>1000000000</max_value><current_value>1</current_value></usage_report><usage_report metric="b36b6076-4ba4-4b6c-b114-d94bf6c8f1d0" period="day"><period_start>2023-04-19 00:00:00 +0000</period_start><period_end>2023-04-20 00:00:00 +0000</period_end><max_value>1000000000</max_value><current_value>0</current_value></usage_report><usage_report metric="b36b6076-4ba4-4b6c-b114-d94bf6c8f1d0" period="hour"><period_start>2023-04-19 09:00:00* Connection #0 to host 127.0.0.1 left intact
 +0000</period_start><period_end>2023-04-19 10:00:00 +0000</period_end><max_value>1000000000</max_value><current_value>0</current_value></usage_report></usage_reports></status>
```
* The worker should report
```
{"severity":"INFO","pid":1,"time":"2023-04-19 09:13:47 UTC","job_class":"ReportJob","runtime":0.00495,"run_plus_queued_time":0.00623,"memoizer_size":8,"memoizer_count":8,"memoizer_hits":0}
```
* The listener should report
```
127.0.0.1 - - [19/Apr/2023 09:13:47 UTC] "GET /transactions/authrep.xml?provider_key=29f5c9c6-b17b-4245-a12d-7eb52705ad8f&service_id=b92c4556-16c3-48a9-85c3-236060e13a56&app_id=8c0211817bddd50b&app_key=a3d8fe2a4e61d521&usage%5Ba5aa2153-6306-4053-b63f-26f1af4ec754%5D=1 HTTP/1.1" 200 2224 0.007327759 0 0 0 618 664 16 - -
```
